### PR TITLE
having a flag to disable update from schedule source

### PR DESCRIPTION
**Pull request by honnix**
@axelri @ulzha

### CODE REVIEW

**Comment by honnix** [comment]
Not super happy with https://github.com/spotify/styx/pull/245/files#diff-a660bb778e38936e1c4f442324f41e64R140

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java [hunk 1]
```diff
@@ -570,6 +571,15 @@ private void setupMetrics(
           workflow.id(),
           () -> stateManager.getActiveStatesCount(workflow.id()));
 
+      final Optional<Workflow> existingWorkflow = cache.workflow(workflow.id());
+      if (existingWorkflow.isPresent()) {
+        // if we have ever registered/updated the workflow from API,
+        // we ignore any subsequent updates via schedule source
+        if (existingWorkflow.get().fromApi().isPresent() && !workflow.fromApi().isPresent()) {
```
Comment (axelri) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: This line reads a bit hard if you're not familiar with the history of this field. What you want to say is, "was this call made through the new API?", but I would read this as "was this call made through an unknown API version?". Ideally we would want to easily compare API versions, both to clarify intent and for future-proofing. For example:

```java
if (isHigher(existingWorkflow.get().fromApi(), workflow.fromApi())) {
  return;
}
```

File: styx-common/src/main/java/com/spotify/styx/model/Workflow.java [hunk 2]
```diff
@@ -41,6 +41,11 @@
   @JsonProperty
   public abstract Optional<URI> componentUri();
 
+  // this gives a hint that the workflow registration/modification came from Styx API
+  // instead of any schedule source
+  @JsonProperty("__from_api")
+  public abstract Optional<String> fromApi();
```
Comment (axelri) [inline styx-common/src/main/java/com/spotify/styx/model/Workflow.java]: String seems like a suboptimal type for the reasons I laid out in my other comment. But maybe it's more convenient for serialization compared to for example Enums.

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java (hunk 1 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: yeah, better to have a properly named method.

File: styx-common/src/main/java/com/spotify/styx/model/Workflow.java (hunk 2 above)
Comment (honnix) [inline styx-common/src/main/java/com/spotify/styx/model/Workflow.java]: good point, as API is part of styx-command anyway, doesn't make sense to go for string.

**Comment by axelri** [comment]
👍

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java [hunk 3]
```diff
@@ -570,11 +571,28 @@ private void setupMetrics(
           workflow.id(),
           () -> stateManager.getActiveStatesCount(workflow.id()));
 
+      final Optional<Workflow> existingWorkflow = cache.workflow(workflow.id());
+      if (existingWorkflow.isPresent()) {
+        if (!isGreaterOrEqualApiVersion(workflow, existingWorkflow.get())) {
+          return;
```
Comment (ulzha) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: Silently returning with a 200 OK? I think that's lying.

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java [hunk 4]
```diff
@@ -570,11 +571,28 @@ private void setupMetrics(
...
+        }
+      }
+
       cache.store(workflow);
       workflowInitializer.inspectChange(workflow);
     };
   }
 
+  // if we have ever registered/updated the workflow via API,
+  // we ignore any subsequent updates via schedule source or lower API version
```
Comment (ulzha) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: This describes "what" but doesn't describe "why". Harrumph. "We ignore _in order to_ enable gradual rollout of workflow configuration API endpoint, while still supporting the polling schedule source for users that haven't migrated".

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java (hunk 3 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: Indeed. Shall we address this later when validating a workflow, e.g. cron syntax? We would need to introduce a new exception and throw it out.

I will anyway put FIXME here.

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java (hunk 4 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: Sure.

**Pull request merged**